python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -445,29 +470,52 @@ def pending_to_running(self, ready_ref: ray.ObjectRef) -> bool:
             return False
         actor = self._pending_actors.pop(ready_ref)
         self._num_tasks_in_flight[actor] = 0
+        self._actor_locations[actor] = ray.get(ready_ref)
         return True
 
-    def pick_actor(self) -> Optional[ray.actor.ActorHandle]:
-        """Provides the least heavily loaded running actor in the pool for task
-        submission.
+    def pick_actor(
+        self, locality_hint: Optional[RefBundle] = None
+    ) -> Optional[ray.actor.ActorHandle]:
+        """Picks an actor for task submission based on busyness and locality.
 
         None will be returned if all actors are either at capacity (according to
         max_tasks_in_flight) or are still pending.
+
+        Args:
+            locality_hint: Try to pick an actor that is local for this bundle.
         """
         if not self._num_tasks_in_flight:
             # Actor pool is empty or all actors are still pending.
             return None
 
-        actor = min(
-            self._num_tasks_in_flight.keys(),
-            key=lambda actor: self._num_tasks_in_flight[actor],
-        )
+        if locality_hint:
+            preferred_loc = self._get_location(locality_hint)
+        else:
+            preferred_loc = None
+
+        def penalty_key(actor):
+            """Returns the key that should be minimized for the best actor.
+
+            We prioritize valid actors, those with argument locality, and those that
+            are not busy, in that order.
+            """
+            busyness = self._num_tasks_in_flight[actor]
+            invalid = busyness >= self._max_tasks_in_flight
```

Comment:
Another thing we should consider in a follow-up: `max_tasks_in_flight` is geared towards pipelining the execution of the current task with the fetching of input data for the next task(s); now that we're taking data locality into account, where we might know that an actor will have most input data local for a particular task and therefore won't have any input data fetching to pipeline with the current task's execution, we might want to rethink how this `max_tasks_in_flight` cap factors in.

This might all be rendered moot once this `max_tasks_in_flight` config is absorbed into the configured `prefetch_batches` arg, when we do the async prefetching/local pipelining port for this actor pool.